metabase/metabase
# Add transient props to styled components

related to [RFC 67](https://www.notion.so/metabase/67-Use-transient-options-with-styled-bb9db73d8e79475c875d384ef5249163)

### Description

Aim of this PR is to add `transient props` like way of working with styled components. We utilize `shouldForwardProp` method and add default value - every property which starts with `$` sign will not be forwarded further

- adds a utility function which doesn't allow passing props which start with `$` to the DOM in styled components 71c4e3c49c55594c40b315350484c1c2e98b9d6c
- replaces all the imports in styled components from `import styled from "@emotion/styled"` to `import { styled } from "metabase/ui/utils"` 4ac1d811cf0023d9051edcb3fc2da06ae51f314b
- adds eslint rule to forbid using `import styled from "@emotion/styled"` in favor of created utility 68ba2c715d6b7861e9a5b67a5f2e86ee42c9b5b2
- fixes all warnings in dev mode regarding passing not recognized props to DOM nodes on the home page ee59b2908ce70bda53e5960b103b3438a06d9ae2
- updates docs a331a4e0da60155f5ea017b8f09d36a0b0e276c2

### How to verify

It should be easier to review commit by commit

Run the app in dev mode, open home page, check browser console - there shouldn't be warnings like

> React does not recognize the `isExpanded` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `isexpanded` instead. If you accidentally passed it from a parent component, remove it from the DOM element

### Demo

| Before | After |
|--------|--------|
| <img width="1889" alt="Screenshot 2023-06-27 at 16 59 24" src="https://github.com/metabase/metabase/assets/125459446/50c38d51-12c5-4674-beec-4a13d1e72071"> | <img width="1906" alt="Screenshot 2023-06-27 at 16 59 41" src="https://github.com/metabase/metabase/assets/125459446/f7af9cb7-7db5-4171-a716-ce00012dc755"> | 

### Checklist

~- [ ] Tests have been added/updated to cover changes in this PR~
